### PR TITLE
010: thread spec

### PR DESCRIPTION
**Pull request by mixmix**

### CODE REVIEW

**Comment by mixmix** [comment]
Preview HERE: https://github.com/ssbc/sips/blob/964e9a5cf407f1adf6e26f3b656c0e769f21caf4/010.md

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
# Threads
```

File: 010.md
Comment (staltz) [inline 010.md:8]:
```suggestion

## Abstract

```

File: 010.md
Comment (staltz) [inline 010.md:28]:
```suggestion
## Terminology

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

## Specification

```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
## Informative references
```

File: 010.md
Comment (staltz) [inline 010.md:459]:
```suggestion
## Reference implementation

- [ssb-threads](https://github.com/ssbc/ssb-threads)

```

File: README.md
Comment (staltz) [inline README.md]:
```suggestion
| 10 | [Threads](010.md) | 2023-03-27 |
```

File: 010.md
Comment (staltz) [inline 010.md:30]:
```suggestion

The following schemas relate to the `content` field in SSB messages. If a field is commented with `// optional`, it means it MAY appear in the `content` field, otherwise it SHOULD appear.

```

**Review by staltz — CHANGES_REQUESTED** [review]
I like the mermaids. :mermaid:

The document needs a little bit of structure, with abstract, specification, etc. Also we need to address those TODOs before merging. And I need to give it a read to make sure it's more prescriptive (telling readers **what** they should do to implement this SIP) than educational (giving readers a tour through threads world), with some MUSTs and SHOULDs.

File: 010.md
Comment (mixmix) [reply 010.md:459]: Need to check with you that you actually do the ordering like this spec suggests
I've seen the linearizing done a couple ways since I've worked on it

I think doing a sort like if A is upstream (in previous chain) of B then -1, if B is upstream of A, then +1, otherwise sort by timestamp?

Comment (staltz) [reply 010.md:459]: ssb-threads uses [ssb-sort](https://github.com/ssbc/ssb-sort/) (which is the reference implementation for the causal/topo sort) so I think it's correct? https://github.com/ssbc/ssb-threads/blob/ca06fcafabf1870bcb29ea24d899235249b7966f/src/index.ts#L279

Comment (mixmix) [reply 010.md:459]: cool. Yeah I remember the logic now, I think it's what I said...
https://github.com/ssbc/ssb-sort/blob/master/index.js#L32-L49

**Comment by mixmix** [comment]
@staltz I want to do the minimal additional work to get this merged.

TODOs:
- I don't know how **forks** are handled, or what the logic is because I didn't implement it. I don't think this should be held up on that
- ~channels ... shrug, I don't have energy, but can point to the right code~

Prescriptive vs. educational, yes keen to see that done.

Given I don't want to do much more work on this, but we need it written down somewhere, would you be down to pair on this to get it done?

File: 010.md
Comment (mixmix) [inline 010.md]:
```suggestion
- SHOULD NOT contain
    - hash symbol `#`
    - punctuation `,."!?`
    - brackets `()[]`
    - spaces ` `, 
- SHOULD NOT be longer than 30 characters long

When parsing a channel, you MUST normalize the channel name by:
1. removing disallowed characters
2. truncating to 30 characters (if required)
3. lowering the case of all characters
    - i.e. `NewZealand` is to be treated the same as `newzealand`, `newZealand` etc.

See [ssb-ref](https://github.com/ssbc/ssb-ref/blob/main/index.js#L95-L104) for a reference implementation of `normalizeChannel`
```

**Comment by staltz** [comment]
> Given I don't want to do much more work on this, but we need it written down somewhere, would you be down to pair on this to get it done?

I'm your inverse, I feel less motivated when I'm pairing to get things done. I suggest we just relax the deadlines with this PR. You (or me) can come back to it some time in the future and finish the TODOs.

**Comment by ahdinosaur** [comment]
> I don't know how forks are handled, or what the logic is because I didn't implement it. I don't think this should be held up on that

for what it's worth, thread forks should probably be a separate spec anyways.

**Comment by mixmix** [comment]
A wild @ahdinosaur appears! Hi <3 
Yeah making it a different sip would feel nice ... or just not requiring it to be in here ...sigh

@staltz can we just merge this. It's better than nothing, and not having it in `master` means it's not really visible

**Comment by staltz** [comment]
> @staltz can we just merge this. It's better than nothing, and not having it in `master` means it's not really visible

Nope! Let's *please* not push documents that have TODOs in them as official specification. If someone really wants to read something in progress, they can just open the "Pull Requests" tab and browse this.

It would also go against the REVIEWING.md which wasn't written in vain, so please don't merge things that are work in progress.

**Comment by mixmix** [comment]
I want to remove forks from this spec, and merge it
Can we do that @staltz ?
I want to be able to link to this at a stable location

**Comment by mixmix** [comment]
DONE - removed forks

**Comment by staltz** [comment]
I'll take a look, thanks

File: 010.md
Comment (staltz) [inline 010.md:19]:
```suggestion

This SIP defines the expected schema for `post` and `vote` messages, and 
how they form a directed acyclic graph known as "threads" in SSB.

```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
  contentWarning: String, // optional
```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
  mentions: Array<Mention>,   // optional
```

I'm suggesting a TypeScript syntax here which should be a lot less ambiguous. One problem in this document is that the syntax `[x]` is ambiguous.

In one case (this case) `[Mention]` means an array of zero, one, or many mentions. But in another case,

```
B => { root: A, previous: [A] }
```

The notation `[A]` meant "an array with only ONE item, `A`".

So let's use `Array<Mention>`

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
  recps: Array<Recp>          // optional
```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
  branch: Array<MsgId>,
```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
  mentions: Array<Mention>,   // optional
```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
  recps: Array<Recp>          // optional
```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
NOTE: sometimes `branch` MAY be a `MsgId` instead of `Array<MsgId>`
```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
  branch: Array<MsgId>      // optional
```

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
Such images SHOULD instead be rendered as a link with an alert/ warning
```

I agree with the intent, but "MUST" is too strong, and it wasn't even used for the `post` msg schema, and it would make Manyverse not SIP10 compliant, which is a bummer. (Yes, Manyverse should be doing what you specced here, PRs welcome)

File: 010.md
Comment (staltz) [inline 010.md]:
```suggestion
When a link to the web is present, the interface SHOULD render an alert to users
```

I don't actually think an alert is important. In Manyverse we use black underscore links for WWW links and blue underscore links for SSB links and that's already one way of signaling to users.

**Review by staltz — CHANGES_REQUESTED** [review]
Super sweet SIP. I like how it's complete: specifies msg schemas, specifies how the tangles works, how channels works, how content warnings work etc.

I have a few suggestions to fix before merging, nothing that requires you to write anything yourself, since I put the code suggestions inline. Mostly, there is some ambiguity with the notation `[X]` which could confuse some readers. That's the most important thing to fix.

Then, there's just a suggestion to relax some MUST to SHOULD.

That's all I have, and then feel free to merge.

File: 010.md
Comment (mixmix) [reply 010.md]: It was too subtle for me to know! Good to know though. Maybe www should be RED

**Comment by mixmix** [comment]
merge squashed!

**Pull request merged**
